programs/stake/src/stake_state.rs

```diff
@@ -757,31 +761,61 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
             match self.state()? {
                 StakeState::Stake(meta, mut stake) => {
                     meta.authorized.check(signers, StakeAuthorize::Staker)?;
+                    let split_rent_exempt_reserve = calculate_split_rent_exempt_reserve(
+                        meta.rent_exempt_reserve,
+                        self.data_len()? as u64,
+                        split.data_len()? as u64,
+                    );
 
                     // verify enough lamports for rent in new stake with the split
-                    if split.lamports()? + lamports < meta.rent_exempt_reserve
-                     // verify enough lamports left in previous stake and not full withdrawal
+                    if split.lamports()? + lamports < split_rent_exempt_reserve
+                    // verify enough lamports left in previous stake and not full withdrawal
                         || (lamports + meta.rent_exempt_reserve > self.lamports()? && lamports != self.lamports()?)
                     {
                         return Err(InstructionError::InsufficientFunds);
                     }
                     // split the stake, subtract rent_exempt_balance unless
-                    //  the destination account already has those lamports
-                    //  in place.
-                    // this could represent a small loss of staked lamports
-                    //  if the split account starts out with a zero balance
-                    let split_stake = stake.split(
-                        lamports - meta.rent_exempt_reserve.saturating_sub(split.lamports()?),
-                    )?;
+                    // the destination account already has those lamports
+                    // in place.
+                    // this means that the new stake account will have a stake equivalent to
+                    // lamports minus rent_exempt_reserve if it starts out with a zero balance
+                    let (remaining_stake_delta, split_stake_amount) = if lamports
+                        == self.lamports()?
+                    {
+                        // If split amount equals the full source stake, the new split stake must
+                        // equal the same amount, regardless of any current lamport balance in the
+                        // split account. Since split accounts retain the state of their source
+                        // account, this prevents any magic activation of stake by prefunding the
+                        // split account.
+                        (
+                            lamports - meta.rent_exempt_reserve,
+                            lamports - split_rent_exempt_reserve,
```

Comment:
Argurably hard to notice, but I think this causes overflow, allowing user to stake around `u64::max_size()` if all of following conditions are met by specifically creating crafted account state and transaction by malicious users:

1. `lamports` is identical as `source.lamports()`.
2. `split` account is bigger than source; so `split_rent_exempt_reserve` > `meta.rent_exempt_reserve`
3. `split` account is pre-funded so that `split.lamports()? + lamports >= split_rent_exempt_reserve`.

Reply:
I think the best thing we could here is to reject such a case? Also, I'd like to add a test for this. (I know you already adding very extensive set of tests. Sorry for more trouble).

Reply:
I think I see. I'll write a test to confirm. Yes, I think rejecting makes sense.

Reply:
Please see `test_split_100_percent_of_source_to_larger_account_edge_case()`